sixtyfps_compiler/llr/lower_expression.rs

```diff
@@ -0,0 +1,977 @@
+// Copyright © SixtyFPS GmbH <info@sixtyfps.io>
+// SPDX-License-Identifier: (GPL-3.0-only OR LicenseRef-SixtyFPS-commercial)
+
+use std::cell::RefCell;
+use std::collections::{BTreeMap, HashMap};
+use std::num::NonZeroUsize;
+use std::rc::{Rc, Weak};
+
+use itertools::Either;
+
+use super::lower_to_item_tree::{LoweredElement, LoweredSubComponentMapping, LoweringState};
+use super::{Animation, PropertyReference};
+use crate::expression_tree::{BuiltinFunction, Expression as tree_Expression};
+use crate::langtype::{EnumerationValue, Type};
+use crate::layout::Orientation;
+use crate::llr::Expression as llr_Expression;
+use crate::namedreference::NamedReference;
+use crate::object_tree::{Element, ElementRc, PropertyAnimation};
+
+pub struct ExpressionContext<'a> {
+    pub component: &'a Rc<crate::object_tree::Component>,
+    pub mapping: &'a LoweredSubComponentMapping,
+    pub state: &'a LoweringState,
+    pub parent: Option<&'a ExpressionContext<'a>>,
+}
+
+impl ExpressionContext<'_> {
+    pub fn map_property_reference(&self, from: &NamedReference) -> Option<PropertyReference> {
+        let element = from.element();
+        let enclosing = &element.borrow().enclosing_component.upgrade().unwrap();
+        if !enclosing.is_global() {
+            let mut map = self;
+            let mut level = 0;
+            while !Rc::ptr_eq(&enclosing, map.component) {
+                map = map.parent.unwrap();
+                level += 1;
+            }
+            if let Some(level) = NonZeroUsize::new(level) {
+                return Some(PropertyReference::InParent {
+                    level,
+                    parent_reference: Box::new(
+                        map.mapping.map_property_reference(from, self.state)?,
+                    ),
+                });
+            }
+        }
+        self.mapping.map_property_reference(from, self.state)
+    }
+}
+
+impl super::EvaluationContext for ExpressionContext<'_> {
+    fn property_ty(&self, _: &PropertyReference) -> &Type {
+        todo!()
+    }
+}
```

Comment:
Not for this merge request, but later: I think this is unreachable code, and with that the only reason for EvaluationContext being a trait. In the cpp branch maybe we can remove that.